File: dust.py
#!/usr/bin/python3
# -*- coding: UTF-8 -*-
# this code indented with actual 0x09 tabs

# This is a hacked up version of code from: https://gist.github.com/geoffwatts/b0b488b5a5257223ed53

import os, sys, serial, time, struct

hostname = os.uname()[1]

home_dir='/home/ghz'

if hostname == 'keen':
	home_dir = '/import/home/ghz'

sys.path.append(home_dir+'/repos/wxlib')
import wxlib as wx

wx_dir = home_dir+'/dust'
plot_d = wx_dir+'/plots/'

def gen_index(pm25, pm10):
	plate = wx_dir+"/dust_wx_index.html.template"
	plate_fd = open(plate, 'r')
	plate_dat = plate_fd.read()
	plate_fd.close()

	ts = time.strftime("%FT%TZ", time.gmtime())

	plate_dat = plate_dat.replace("TTTPM25", str("%.2f" % pm25))
	plate_dat = plate_dat.replace("TTTPM10", str("%.2f" % pm10))
	plate_dat = plate_dat.replace("DATE", ts)

	wx.write_out(wx_dir+'/plots/dust_wx.html', plate_dat, 'w')

if __name__ == "__main__":

	dat_fname = 'dust.dat'
	wx.proof_dir(plot_d)

	ser = serial.Serial()
	# ser.port = "/dev/ttyU0"
	ser.port = "/dev/ttyUSB0"
	ser.baudrate = 9600

	ser.open()
	ser.flushInput()

	byte, lastbyte = b"\x00", b"\x00"

	debug = 1
	time0 = time1 = time.time()
	pm_25_val = pm_10_val = count = 0

	if debug:
		print("entering loop", flush=True)

	while True:
		lastbyte = byte
		byte = ser.read(size=1)
    
		if debug:
			print("0", end="", flush=True)

		# We got a valid packet header
		if lastbyte == b"\xAA" and byte == b"\xC0":
			sentence = ser.read(size=8) # Read 8 more bytes

			if debug:
				print("1", end="", flush=True)

			check = b"\x00\x00\x00\x00\x00\x00\x00\x00"
			check = struct.unpack('<cccccccc', sentence)

			if debug:
				print("2", end="", flush=True)

			# check tail byte, should also help eliminate short reads
			if check[7] != b"\xAB":
				print("tail byte not 0xAB", flush=True)
				continue

			if debug:
				print("3", end="", flush=True)

			# if it worked move on to checksum
			b = bytearray(sentence[0:7])
			if b[6] != (sum(b[0:6]) % 256):
				print("checksum failed", flush=True)
				continue

			if debug:
				print("4", end="", flush=True)

			# Decode the packet - little endian, 2 shorts for pm2.5 and pm10, 2 ID bytes, checksum, message tail
			readings = struct.unpack('<hhxxcc',sentence)
        
			pm_25_val += readings[0]
			pm_10_val += readings[1]

			if debug:
				print(".", end="", flush=True)

			count += 1
			time1 = time.time()
			if((time1 - time0) > 60):
				ts = time.strftime("%FT%T%Z", time.gmtime(float(time1)))
				tsf = time.strftime("%Y%m%d%H%M%S", time.gmtime(float(time1)))
				pm_25 = pm_25_val / count / 10.0
				pm_10 = pm_10_val / count / 10.0
				dat_string = "%s\tPM 2.5: %.2f μg/m³\tPM 10: %.2f μg/m³\n" % (ts, pm_25, pm_10)
				wx.write_out_dat_stamp(tsf, dat_fname, dat_string, wx_dir)

				if debug:
					print("\n" + dat_string, end="", flush=True)

				if hostname == 'elf':
					gen_index(pm_25, pm_10)

				pm_25_val = pm_10_val = count = 0
				time0 = time1 = time.time()

File: dust.pl
#!/usr/bin/perl

use strict;
use warnings;
use utf8;
use open qw(:std :utf8);
use Encode qw(encode decode);
use POSIX qw(strftime);
use Device::SerialPort;

sub usage() {
	print "usage: $0 serial_port\n";
	exit;
}

my $data_dir_t = "/import/home/ghz/repos/dust_wx/data";

sub write_out_dat($$) {
	my ($pm25, $pm10) = @_;
	my $f_tplate = "dust.dat";
	my $date_y = strftime("%Y", gmtime);
	my $tdate = strftime("%Y%m%d", gmtime);
	my $ts = strftime("%FT%T%Z", gmtime);
	my $data_dir = "$data_dir_t/$date_y";

	unless ( -d $data_dir) {
		mkdir $data_dir or die "couldn't make $data_dir";
	}

	my $dat_f = "$data_dir/$f_tplate.$tdate";
	open(OUT, ">>", $dat_f) or die "omg! can't open output file: $dat_f";
	printf OUT "%s\tPM 2.5: %.2f µ/m³\tPM 10: %.2f µ/m³\n", $ts, $pm25 / 10.0, $pm10 / 10.0;
	close OUT;
}

unless ($#ARGV == 0) {
	usage();
}

@ARGV = map { decode("UTF-8", $_) } @ARGV;
my $read_f = $ARGV[0];

unless (-c  $read_f) {
	usage;
}

unless ( -d $data_dir_t) {
	print "$data_dir_t doesn't exit\n";
	mkdir $data_dir_t or die "couldn't make $data_dir_t";
}

my $port = Device::SerialPort->new($read_f) || die "serial port open failed";

# 9600n81
$port->baudrate(9600);
$port->parity("none");
$port->databits(8);
$port->stopbits(1);

$port->purge_all;
# $port->purge_rx;
# $port->purge_tx;

# from the docs: https://metacpan.org/pod/Device::SerialPort
# keeps my load astoundingly lower
$port->read_const_time(100);	# const time for read (milliseconds)
$port->read_char_time(5);	# avg time between read char

my $debug = 0;
my ($itr, $pm25_t, $pm10_t) = (0, 0, 0);
my($b0, $ub, $cnt) = (hex("0xde"), hex("0xad"), hex("0xbe"));

while (1) {
	($cnt, $b0) = $port->read(1);
	$ub = unpack('C', $b0);	
	printf "ub: 0x%02x\t", $ub if $ub && $debug;
	printf "cnt: %d\n", $cnt if $cnt && $debug;
	if ($ub and $ub == hex("0xaa")) {
		($cnt, $b0) = $port->read(1);
		$ub = unpack('C', $b0);	
		if ($ub == hex("0xc0")) {
			($cnt, $b0) = $port->read(8);
			my ($c0, $c1, $c2, $c3, $c4, $c5, $c6, $c7) = unpack('CCCCCCCC', $b0);	
			if ($c7 != hex("0xab")) {
				printf "tail byte fail! 0x%02x != 0xab\n", $c7;
				next;
			}

			my $csum = $c0 + $c1 + $c2 + $c3 + $c4 + $c5;
			if ($csum % 256 != $c6){
				printf "csum mod 256: 0x%02x, checksum: 0x%02x\n", $csum % 256, $c6;
				next;
			}

			my($pm25, $pm10, $b1, $b2, $b3, $b4) = unpack('vvCCCC', $b0);	
			printf "%.2f, %.2f, 0x%02x, 0x%02x, 0x%02x, 0x%02x, 0x%02x\n",
				$pm25 / 10.0, $pm10 / 10.0, $b1, $b2, $b3, $b4, $csum % 256 if $debug == 1;

			$pm25_t += $pm25;
			$pm10_t += $pm10;
			$itr++;
			if ($itr >= 60) {
				write_out_dat($pm25_t/$itr/10.0, $pm10_t/$itr/10.0);
				$pm25_t = 0;
				$pm10_t = 0;
				$itr = 0;
			}
		}
	}
}
